presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java

```diff
@@ -304,128 +290,146 @@ public void addInput(Page page)
+                    return needsMoreData();
+                }
+
+                partition.processNextRow(pageBuilder);
             }
-        }
-        else {
-            // We had previous results buffered, but the new page starts with new group values
-            return page;
+
+            Page page = pageBuilder.build();
+            pageBuilder.reset();
+            return ofResult(page, !partition.hasNext());
         }
     }
 
-    private static Page rearrangePage(Page page, int[] channels)
+    private class ProduceWindowPartitions
+            implements Function<PagesIndex, WorkProcessor<WindowPartition>>
     {
-        Block[] newBlocks = new Block[channels.length];
-        for (int i = 0; i < channels.length; i++) {
-            newBlocks[i] = page.getBlock(channels[i]);
+        @Override
+        public WorkProcessor<WindowPartition> apply(PagesIndex pagesIndex)
+        {
+            return WorkProcessor.create(new WorkProcessor.Process<WindowPartition>()
+            {
+                int partitionStart;
+
+                @Override
+                public WorkProcessor.ProcessorState<WindowPartition> process()
+                {
+                    if (partitionStart == pagesIndex.getPositionCount()) {
+                        return finished();
+                    }
+
+                    int partitionEnd = findGroupEnd(pagesIndex, unGroupedPartitionHashStrategy, partitionStart);
+                    WindowPartition partition = new WindowPartition(pagesIndex, partitionStart, partitionEnd, outputChannels, windowFunctions, peerGroupHashStrategy);
+                    windowInfo.addPartition(partition);
+                    partitionStart = partitionEnd;
+                    return ofResult(partition);
+                }
+            });
         }
-        return new Page(page.getPositionCount(), newBlocks);
     }
 
-    @Override
-    public Page getOutput()
+    private class ProducePagesIndexes
+            implements WorkProcessor.Process<PagesIndex>
     {
-        if (state == State.NEEDS_INPUT || state == State.FINISHED) {
-            return null;
+        private final LocalMemoryContext memoryContext;
+        private boolean resetPagesIndex;
+
+        private ProducePagesIndexes(AggregatedMemoryContext memoryContext)
+        {
+            this.memoryContext = memoryContext.newLocalMemoryContext(ProducePagesIndexes.class.getSimpleName());
         }
 
-        Page page = extractOutput();
-        localUserMemoryContext.setBytes(pagesIndex.getEstimatedSize().toBytes());
-        return page;
-    }
+        @Override
+        public WorkProcessor.ProcessorState<PagesIndex> process()
+        {
+            if (resetPagesIndex) {
+                pagesIndex.clear();
+                updateMemoryUsage();
+                resetPagesIndex = false;
+            }
 
-    private Page extractOutput()
-    {
-        // INVARIANT: pagesIndex contains the full grouped & sorted data for one or more partitions
```

Comment:
would be good to retain this comment inside `com.facebook.presto.operator.WindowOperator.ProduceWindowPartitions#apply` or somewhere near